modules/internal/Atomics.chpl

```diff
@@ -278,148 +101,205 @@ module Atomics {
     chpl_rmem_consist_fence(order);
   }
 
+
+  private proc isSupported(type T) param {
+    return T == bool || isInt(T) || isUint(T) || isReal(T);
+  }
+
+  // Compute the C/Runtime type from the Chapel type
+  // TODO support extern type renaming?
+  private proc externT(type T) type {
+    extern type atomic_bool;
+
+    extern type atomic_int_least8_t;
+    extern type atomic_int_least16_t;
+    extern type atomic_int_least32_t;
+    extern type atomic_int_least64_t;
+
+    extern type atomic_uint_least8_t;
+    extern type atomic_uint_least16_t;
+    extern type atomic_uint_least32_t;
+    extern type atomic_uint_least64_t;
+
+    extern type atomic__real64;
+    extern type atomic__real32;
+
+    select T {
+      when bool     do return atomic_bool;
+
+      when int(8)   do return atomic_int_least8_t;
+      when int(16)  do return atomic_int_least16_t;
+      when int(32)  do return atomic_int_least32_t;
+      when int(64)  do return atomic_int_least64_t;
+
+      when uint(8)  do return atomic_uint_least8_t;
+      when uint(16) do return atomic_uint_least16_t;
+      when uint(32) do return atomic_uint_least32_t;
+      when uint(64) do return atomic_uint_least64_t;
+
+      when real(32) do return atomic__real32;
+      when real(64) do return atomic__real64;
+    }
```

Comment:
For sure no need for an `otherwise` here, to guard against future mis-coding?

Reply:
Pretty sure you'd get a compilation error in that case because the function wouldn't return anything. The `select T` has to be evaluated at compile-time since this is a type function.

Reply:
Right, it will return void -- and these functions are all indirectly guarded by `chpl__processorAtomicType()`, which issues a compileError for unsupported types

Reply:
Isn't that moot, since it already doesn't return anything if the type doesn't match any of the `when` options?  I was thinking of putting a `compilerError()` in the `otherwise`, the moral equivalent of what is done at https://github.com/chapel-lang/chapel/pull/10588/files#diff-b8fb835d4f7a58ecdc2589f832485f7aR161.  We'll get some kind of compiler error either way, but a `compilerError()` in an `otherwise` allows that to be a more specific one.

Reply:
Right, we could add compilerErrors to `externT()`, `externTString()`, and a few other routines but they're already guarded by the compilerError in `chpl__processorAtomicType()` and I didn't see the need to duplicate the error.